teps/0114-custom-tasks-beta.md

```diff
@@ -0,0 +1,342 @@
+the API such that it's `Run.Spec.CustomSpec`.
+
+  ```yaml
+  # before
+  apiVersion: tekton.dev/v1alpha1
+  kind: Run
+  metadata:
+    name: run-with-spec
+  spec:
+    spec:
+      apiVersion: example.dev/v1alpha1
+      kind: Example
+      spec:
+        field1: value1
+        field2: value2
+  # after
+  apiVersion: tekton.dev/v1beta1
+  kind: CustomRun
+  metadata:
+    name: customrun-with-spec
+  spec:
+    customSpec:
+      apiVersion: example.dev/v1beta1
+      kind: Example
+      spec:
+        field1: value1
+        field2: value2
+  ```
+
+Today, [references][runs-spec-docs] are in `Run.Spec.Ref`. We propose aligning the embedded specs with the rest of the
+API such that it's `Run.Spec.CustomRef`.
+
+  ```yaml
+  # before
+  apiVersion: tekton.dev/v1alpha1
+  kind: Run
+  metadata:
+    name: run-with-reference
+  spec:
+    ref:
+      apiVersion: example.dev/v1alpha1
+      kind: Example
+      name: my-example
+  ---
+  # after
+  apiVersion: tekton.dev/v1beta1
+  kind: CustomRun
+  metadata:
+    name: customrun-with-reference
+  spec:
+    customRef:
+      apiVersion: example.dev/v1beta1
+      kind: Example
+      name: my-customtask
+  ```
+
+Note that the API for `Tasks` and `Pipelines` is:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: pipelinerun-with-ref
+spec:
+  pipelineRef:
+    name: my-pipeline
+---
+apiVersion: tekton.dev/v1beta1
+kind: TaskRun
+metadata:
+  name: taskrun-with-reference
+spec:
+  taskRef:
+    name: my-task
+---
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: pipelinerun-with-spec
+spec:
+  pipelineSpec:
+    params:
+      - name: param1
+    tasks:
+      - name: my-task
+---
+apiVersion: tekton.dev/v1beta1
+kind: TaskRun
+metadata:
+  name: taskrun-with-spec
+spec:
+  taskSpec:
+    params:
+      - name: param1
+    steps:
+      - name: my-step
+```
+
+Simplifying all the references to use `ref` and embedded specifications to use `spec` is out of scope for this TEP. That
+work is tracked in [tektoncd/pipeline#5138][5138].
+
+##### Feature Gates
+
+Remove guarding of `Custom Tasks` behind `enable-custom-tasks` and `enable-api-fields` feature gates.
+
+When [TEP-0096: Pipelines V1 API][tep-0096] is implemented to add V1 API, `Custom Tasks` will be gated behind feature
+gate `enable-api-fields` being set to `"beta"` - this is out of scope for this TEP  (in scope for TEP-0096).
+
+#### Documentation
+
+Expand the documentation for [`Custom Tasks`][custom-tasks-docs] and [`Runs`][runs-docs]. 
+
+#### Testing
+
+Improve the test coverage for `Custom Tasks` and `Runs`. 
+
+Writing end-to-end tests for `Custom Tasks` and `Runs` has been difficult because we don't have `Custom Tasks` 
+controllers in Tekton Pipelines. We propose adding a simple `Custom Task` controller in Tekton Pipelines that's used 
+only for tests. This could be a version of the [`Wait Custom Task`][wait] that simply waits a given amount of time 
```

Comment:
I dont have any particular preference here but `cel` might be a little less lighter compared to `wait` if you want to consider it ...

Reply:
added `cel` as an option too - we can leave that choice to the implementation details